### ISSUE TITLE
Improve the way we handle CSS to accomodate for the growing complexity of the project

### ISSUE THREAD

**Issue by Skaiir** [bpmn-io/form-js#257]
__Is your feature request related to a problem? Please describe__

As our component library grows the CSS files used to describe them are starting to become more complicated and harder to read. It would be good for us to start using CSS pre-processors instead.

__Describe the solution you'd like__

Adopt a CSS preprocessor like LESS and integrate it all into our CI.

__Describe alternatives you've considered__

n/a

__Additional context__

n/a

**Comment by MaxTru** [comment]
Why not use LESS (e.g., we also use it in camunda-modeler and plan to use it for propPanel: https://github.com/bpmn-io/properties-panel/issues/48)

**Comment by pinussilvestrus** [comment]
My two cents from a dev perspective: I personally never had big issues from working with a single "vanilla" CSS file. I experienced many headaches from using pre-processors as they add more magic to your bundling pipeline. :sweat_smile: Most of the time, I was just happy to use my well-known CSS.

Things, as we have in the Camunda Modeler to split styles, are cool to keep things focused. Having a central place would be fine, too. I have no strong feelings about this. 

By the way, a cool thing I like [about Svelte](https://svelte.dev/tutorial/styling); having the styles in the same file as the component, and then the compiler simply bundles your CSS together. 

I think the styles in our libraries still have a reasonable size. That might change in the future, though, but maybe then a better question would be whether to split the libraries, not the styles.

**Comment by marstamm** [comment]
Some thoughts on this one:
**As a Developer**
- I like using vanilla CSS, I know what the selectors do and don't have to remember what special operators do in sass or less.
- When I was refactoring Drilldown CSS, I missed some class names in the big CSS files. It is not always clear from the names which style is required by which component.
- It would be great to have Code spliting for CSS, so I can easily see what styles belong to the component and makes refactoring easier. We could use a [rollup plugin](https://www.npmjs.com/package/rollup-plugin-import-css) to include them in JS and bundle them all in a single CSS file for the distro.
- This can also result in smaller CSS files for viewers, if they don't have to include all modeler styles.

**As a library user**
- I don't want to care about CSS of the libraries I import. I expect them to work out of the box.
- Tools like codemirror inline their styles and don't require you to include a CSS file in you app.
- With a quick google search, there are processors that do that for you, like [ESifyCSS](https://github.com/gjbkz/esifycss) (also available as a [rollup plugin](https://www.npmjs.com/package/rollup-plugin-embed-css)). This will most likely be out of Scope for this one

**TL;DR:** I like this initiative and think we should change our handling of CSS, but make it as easy as possible for everyone involved

**Comment by Skaiir** [comment]
**General:**

- If we adopt any kind of splitting of css files, we should 100% use rollup to bundle them all together, as multiple .css files results in multiple HTML calls which is not something desirable.
- Team is relatively split (ha) on the idea of splitting the files, but more pro than con.
- The implementation cost for this is quite low, I think I could set it up in an hour.
- Team is also split on pre-processors, but more con than pro.


**The costs and value of switching over to split files:**

Costs: 
- Initial setup cost of rollup configuration for bundling (xs)
- Splitting out all the files will cause some conflicts in my current work and will require revision (xs)
- Git history will have one major change

Value:
- Keeping css close to our js files, making modifications to css quicker from a development standpoint
- In the case of form-js, there are plans to expand on the component library in the future and this will solve the scaling problem early
- Enabling easier refactoring, if we ever decide to extract the components outside of the library for some reason we don't need to surgically remove the css from one big file, moreover it allows us essentially copy paste that component with its styling into any other library that does css this way. 

_Fun fact, we already use a bundler for the form-js playground (doesn't do that much but it's there :D)_


**How do we consume it?**

Switching over to a rollup plugin for bundling the css would require us to include the css files in a js file somewhere. Note that this does not mean the CSS is scoped to that React component, it only really 'tags' it to be bundled up. But that's fine.

Using less and scss there's also the possibility to use @imports and partials to achieve the same behavior. That is a little more complex to do because it requires referencing all those partials and aggregating them by writing out all the individual paths, something that the hierarchy of .js files already naturally provides in the rollup plugin case. Even if we were using preprocessors or decided to switch over to them in the future, I would still recommend the first approach of bundling naively without using special scss / less features. 


**About scss / less**

Firstly _let's forget about sass_ because the syntax is more like yaml which we're quite far from.
Doing some research, I've concluded that the differences between those two are for our cases negligible. The feature set is pretty much 1 to 1 with some exceptions for procedural css generation which we'll honestly never touch. Although overall scss has more features. In terms of usage scss trumps less quite a bit. Both are open source. Less is also js based while sass is ruby based. 

One place where they do differ is (at least for rollup), the less plugin has barely any usage, while the scss one is well maintained and has many downloads.

I would say for us, there's probably no practical difference. Scss has a slight edge on popularity and will most likely slowly drown out less, but at the same time it's not likely that we'll ever need features beyond the basic ones that have existed within both since pretty much their inception. 

However when it comes to community contributors, they will be more likely to be familiar with scss than less. So if we do decide to switch over, I think scss is the right choice. Probably makes very little measurable difference in the end but it's slightly safer. 


**Do we even need a preprocessor?**

I think this depends on the complexity of our components.

Here's an example of css vs scss for the taglist component I'm building. Note that this is the most complex component to date in form-js and I couldn't find one more biased towards using a preprocessor. If we're not sure about the value here, then there's no value across the codebase. 

**CSS**
```scss
.fjs-container .fjs-taglist {
	display: flex;
	flex-wrap: wrap;
	gap: 5px;
	border: var(--color-borders) solid 1px;
	border-radius: 3px;
	padding: 5px;
	background-color: var(--color-background);
}

.fjs-container .fjs-taglist:focus-within {
	outline: var(--color-borders) solid 1px;
}

.fjs-container .fjs-taglist.disabled {
	border: var(--color-borders-disabled) solid 1px;
	background-color: var(--color-background-disabled);
}

.fjs-container .fjs-taglist .fjs-taglist-tag {
	display: flex;
	overflow: hidden;
	user-select: none;
	border-radius: 2px;
	background-color: var(--color-grey-225-10-90);
}

.fjs-container .fjs-taglist .fjs-taglist-tag .fjs-taglist-tag-label {
	padding: 2px 6px 2px 8px;
}

.fjs-container .fjs-taglist .fjs-taglist-tag .fjs-taglist-tag-remove {
	overflow: clip;
	width: 22px;
	height: 22px;
	text-align: center;
	line-height: 32px;
	background-color: var(--color-grey-225-10-75);
}

.fjs-container .fjs-taglist .fjs-taglist-tag .fjs-taglist-tag-remove:hover {
	background-color: var(--color-grey-225-10-55);
}

.fjs-container .fjs-taglist .fjs-taglist-input {
	border: none;
	background-color: transparent;
	min-width: 100px;
	height: 22px;
	flex-grow: 1;
}

.fjs-container .fjs-taglist .fjs-taglist-input:focus-visible {
  outline: none;
}

.fjs-container .fjs-taglist .fjs-taglist-dropdown-anchor {
	position: relative;
}

.fjs-container .fjs-dropdownlist {
	user-select: none;
	overflow-y: auto;
	scroll-behavior: smooth;
	width: 100%;
	border-radius: 3px;
	margin-top: 3px;
	box-shadow: 0px 0px 5px var(--color-grey-225-10-85);
	background-color: var(--color-white);
}

.fjs-container .fjs-dropdownlist .fjs-dropdownlist-item {
	padding: 4px 8px;
	border-bottom: 1px solid var(--color-grey-225-10-85);
	color: var(--color-grey-225-10-35);
}

.fjs-container .fjs-dropdownlist .fjs-dropdownlist-item:last-child {
	border-bottom: none;
}

.fjs-container .fjs-dropdownlist .fjs-dropdownlist-item.focused {
	background-color: var(--color-grey-225-10-85);
	color: var(--color-black);
}

.fjs-container .fjs-dropdownlist .fjs-dropdownlist-empty {
	padding: 4px 8px;
	color: var(--color-text-lighter);
}
```
**SCSS** (I rushed the conversion using a tool)
```scss
.fjs-container {

	.fjs-taglist {
		display: flex;
		flex-wrap: wrap;
		gap: 5px;
		border: var(--color-borders) solid 1px;
		border-radius: 3px;
		padding: 5px;
		background-color: var(--color-background);
		
		&:focus-within {
			outline: var(--color-borders) solid 1px;
		}
		
		&.disabled {
			border: var(--color-borders-disabled) solid 1px;
			background-color: var(--color-background-disabled);
		}
		
		.fjs-taglist-tag {
			display: flex;
			overflow: hidden;
			user-select: none;
			border-radius: 2px;
			background-color: var(--color-grey-225-10-90);
			
			.fjs-taglist-tag-label {
				padding: 2px 6px 2px 8px;
			}
			
			.fjs-taglist-tag-remove {
				overflow: clip;
				width: 22px;
				height: 22px;
				text-align: center;
				line-height: 32px;
				background-color: var(--color-grey-225-10-75);
				&:hover {
					background-color: var(--color-grey-225-10-55);
				}
			}
		}
		
		.fjs-taglist-input {
			border: none;
			background-color: transparent;
			min-width: 100px;
			height: 22px;
			flex-grow: 1;
			&:focus-visible {
				outline: none;
			}
		}
		
		.fjs-taglist-dropdown-anchor {
			position: relative;
		}
	}

	.fjs-dropdownlist {
		user-select: none;
		overflow-y: auto;
		scroll-behavior: smooth;
		width: 100%;
		border-radius: 3px;
		margin-top: 3px;
		box-shadow: 0px 0px 5px var(--color-grey-225-10-85);
		background-color: var(--color-white);
		
		.fjs-dropdownlist-item {
			padding: 4px 8px;
			border-bottom: 1px solid var(--color-grey-225-10-85);
			color: var(--color-grey-225-10-35);
			
			&:last-child {
				border-bottom: none;
			}
		}
		
		.fjs-dropdownlist-item.focused {
			background-color: var(--color-grey-225-10-85);
			color: var(--color-black);
		}
		
		.fjs-dropdownlist-empty {
			padding: 4px 8px;
			color: var(--color-text-lighter);
		}
	}
}
```

IMO the only feature we really care about is the nesting to improve readability. 

Personally I find it just easier to mentally match to the component itself, but I've been working on this for a while now so I'd like external opinions here.

In the end infrastructure wise, there won't be much of a cost to adding scss. But this is the perfect scenario where it actually does help, smaller components won't benefit from it as much. But then again, scss is a superset of css, switching over to it does not stop us from using regular css if that's what we want.

So I have no conclusive decision here I'd like to discuss it just a little more with people. 

**My pitch:**

- We adopt splitting files no matter what, this is a cheap implementation at this point and most of us agree there's some value in good code separation and will make development easier. 
- We still deliberate on preprocessors a little, but the current idea is that we do not include it unless we change our minds as a team. If we do use one, I'd vouch for scss for popularity's sake. IMO the more accessible we can be to external contributors the better.  I think the fact that we used less in the past doesn't stop us here, if it makes sense we can convert that codebase over to scss as well (most likely not a big pain point). 

I'll be on the lookout for further feedback 😄

**Comment by Skaiir** [comment]
@marstamm 

Yes but it won't achieve the same thing. `@import` in plain css leads to multiple html calls while in preprocessors it combines the files into one.

**Comment by pinussilvestrus** [comment]
Nice summary @Skaiir 👏 

Given the CSS vs. SCSS Taglist example, I have to admit that I still like the CSS version more, even readability wise 😄  

> Fun fact, we already use a bundler for the form-js playground (doesn't do that much but it's there :D)

That's exactly the approach I would vision it. Define the styles per component (e.g. [Playground.css](https://github.com/bpmn-io/form-js/blob/master/packages/form-js-playground/src/Playground.css)) and bundle it all together in production [(rollup config](https://github.com/bpmn-io/form-js/blob/master/packages/form-js-playground/rollup.config.js#L22)). The [connectors extension](https://github.com/bpmn-io/bpmn-js-connectors-extension) is doing the same. No real pre-processor,[ just CSS files](https://github.com/bpmn-io/bpmn-js-connectors-extension/blob/master/src/element-template-chooser/ElementTemplateChooser.js#L28).

**Comment by Skaiir** [comment]
I'm gonna shelf this for now before I merge in the new form components.

**Comment by nikku** [comment]
We should carefully consider @marstamm's feedback (https://github.com/bpmn-io/form-js/issues/257#issuecomment-1131483880):

Do we want our users to bother with `CSS` files? Why not inline everything?

This is obviously independent from the CSS splitting aspect.

**Comment by Skaiir** [comment]
Yeah thanks for highlighting this @nikku I glossed over it a little. I'll take this into more consideration whenever I pick the topic up again.
